File: checkers.js
// checkers.js
const readline = require('readline');

const EMPTY = 0;
const WHITE = 1;
const BLACK = 2;

class Checkers {
    constructor() {
        this.board = Array.from({ length: 8 }, () => Array(8).fill(EMPTY));
        for (let row = 0; row < 8; row++) {
            for (let col = 0; col < 8; col++) {
                if ((row + col) % 2 === 1) {
                    if (row < 3) this.board[row][col] = WHITE;
                    else if (row > 4) this.board[row][col] = BLACK;
                }
            }
        }
        this.turn = WHITE;
        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
    }

    printBoard() {
        console.log('  a b c d e f g h');
        for (let row = 0; row < 8; row++) {
            let line = (row + 1) + ' ';
            for (let col = 0; col < 8; col++) {
                const ch = this.board[row][col] === EMPTY ? '·' :
                           this.board[row][col] === WHITE ? '○' : '●';
                line += ch + ' ';
            }
            console.log(line);
        }
    }

    isValidPos(row, col) {
        return row >= 0 && row < 8 && col >= 0 && col < 8;
    }

    getMoves(row, col) {
        const piece = this.board[row][col];
        if (piece === EMPTY) return [];
        const dir = piece === WHITE ? 1 : -1;
        const moves = [];
        // простые ходы
        for (const dc of [-1, 1]) {
            const nr = row + dir, nc = col + dc;
            if (this.isValidPos(nr, nc) && this.board[nr][nc] === EMPTY) {
                moves.push({ from: [row, col], to: [nr, nc], capture: null });
            }
        }
        // захваты
        for (const dc of [-1, 1]) {
            const nr = row + dir*2, nc = col + dc*2;
            if (this.isValidPos(nr, nc)) {
                const mr = row + dir, mc = col + dc;
                if (this.isValidPos(mr, mc) && this.board[mr][mc] !== EMPTY && this.board[mr][mc] !== piece && this.board[nr][nc] === EMPTY) {
                    moves.push({ from: [row, col], to: [nr, nc], capture: [mr, mc] });
                }
            }
        }
        return moves;
    }

    getAllMoves(color) {
        let moves = [];
        for (let row = 0; row < 8; row++) {
            for (let col = 0; col < 8; col++) {
                if (this.board[row][col] === color) {
                    moves = moves.concat(this.getMoves(row, col));
                }
            }
        }
        return moves;
    }

    hasCaptures(color) {
        for (let row = 0; row < 8; row++) {
            for (let col = 0; col < 8; col++) {
                if (this.board[row][col] === color) {
                    for (const m of this.getMoves(row, col)) {
                        if (m.capture) return true;
                    }
                }
            }
        }
        return false;
    }

    makeMove(move) {
        const [fr, fc] = move.from;
        const [tr, tc] = move.to;
        this.board[tr][tc] = this.board[fr][fc];
        this.board[fr][fc] = EMPTY;
        if (move.capture) {
            const [cr, cc] = move.capture;
            this.board[cr][cc] = EMPTY;
        }
    }

    isGameOver() {
        const whiteMoves = this.getAllMoves(WHITE);
        const blackMoves = this.getAllMoves(BLACK);
        if (whiteMoves.length === 0 || blackMoves.length === 0) return true;
        let hasWhite = false, hasBlack = false;
        for (let row = 0; row < 8; row++) {
            for (let col = 0; col < 8; col++) {
                if (this.board[row][col] === WHITE) hasWhite = true;
                else if (this.board[row][col] === BLACK) hasBlack = true;
            }
        }
        return !hasWhite || !hasBlack;
    }

    getWinner() {
        const whiteMoves = this.getAllMoves(WHITE);
        const blackMoves = this.getAllMoves(BLACK);
        let hasWhite = false, hasBlack = false;
        for (let row = 0; row < 8; row++) {
            for (let col = 0; col < 8; col++) {
                if (this.board[row][col] === WHITE) hasWhite = true;
                else if (this.board[row][col] === BLACK) hasBlack = true;
            }
        }
        if (!hasWhite || whiteMoves.length === 0) return BLACK;
        if (!hasBlack || blackMoves.length === 0) return WHITE;
        return 0;
    }

    parseMove(s) {
        const parts = s.trim().split(/\s+/);
        if (parts.length !== 2) return null;
        if (parts[0].length < 2 || parts[1].length < 2) return null;
        const col1 = parts[0].charCodeAt(0) - 97;
        const row1 = parseInt(parts[0][1]) - 1;
        const col2 = parts[1].charCodeAt(0) - 97;
        const row2 = parseInt(parts[1][1]) - 1;
        if (row1 < 0 || row1 > 7 || col1 < 0 || col1 > 7 || row2 < 0 || row2 > 7 || col2 < 0 || col2 > 7) return null;
        return { from: [row1, col1], to: [row2, col2] };
    }

    play() {
        console.log('Добро пожаловать в мини-шашки!');
        console.log('Белые (○) ходят первыми.');
        console.log('Вводите ход в формате: a2 b3');
        const prompt = () => {
            if (this.isGameOver()) {
                this.printBoard();
                const winner = this.getWinner();
                if (winner === WHITE) console.log('Победили белые (○)!');
                else if (winner === BLACK) console.log('Победили чёрные (●)!');
                else console.log('Ничья?');
                this.rl.close();
                return;
            }
            this.printBoard();
            const color = this.turn;
            const colorName = color === WHITE ? 'белых (○)' : 'чёрных (●)';
            console.log(`Ход ${colorName}.`);
            const allMoves = this.getAllMoves(color);
            if (allMoves.length === 0) {
                console.log('Нет доступных ходов, игра окончена.');
                this.rl.close();
                return;
            }
            const hasCap = this.hasCaptures(color);
            let moves = allMoves;
            if (hasCap) {
                console.log('Обязательный захват!');
                moves = allMoves.filter(m => m.capture !== null);
            }
            this.rl.question('> ', (cmd) => {
                if (cmd.trim() === 'quit') { this.rl.close(); return; }
                const parsed = this.parseMove(cmd);
                if (!parsed) {
                    console.log('Неверный формат. Используйте: a2 b3');
                    prompt();
                    return;
                }
                const found = moves.find(m => m.from[0] === parsed.from[0] && m.from[1] === parsed.from[1] &&
                                           m.to[0] === parsed.to[0] && m.to[1] === parsed.to[1]);
                if (!found) {
                    console.log('Неверный ход. Попробуйте снова.');
                    prompt();
                    return;
                }
                this.makeMove(found);
                this.turn = color === WHITE ? BLACK : WHITE;
                prompt();
            });
        };
        prompt();
    }
}

const game = new Checkers();
game.play();

File: checkers.py
# checkers.py
import sys

EMPTY = 0
WHITE = 1  # ходят первыми
BLACK = 2

class Checkers:
    def __init__(self):
        self.board = [[EMPTY]*8 for _ in range(8)]
        # расстановка: белые снизу (строки 0-2), чёрные сверху (строки 5-7)
        for row in range(3):
            for col in range(8):
                if (row + col) % 2 == 1:
                    self.board[row][col] = WHITE
        for row in range(5, 8):
            for col in range(8):
                if (row + col) % 2 == 1:
                    self.board[row][col] = BLACK
        self.turn = WHITE  # белые ходят первыми

    def print_board(self):
        print("  a b c d e f g h")
        for row in range(8):
            print(row+1, end=" ")
            for col in range(8):
                piece = self.board[row][col]
                if piece == EMPTY:
                    ch = '·'
                elif piece == WHITE:
                    ch = '○'
                else:
                    ch = '●'
                print(ch, end=" ")
            print()

    def is_valid_pos(self, row, col):
        return 0 <= row < 8 and 0 <= col < 8

    def get_piece(self, row, col):
        if self.is_valid_pos(row, col):
            return self.board[row][col]
        return None

    # Получить все ходы для шашки (без учёта обязательности)
    def get_moves(self, row, col):
        piece = self.board[row][col]
        if piece == EMPTY:
            return []
        direction = 1 if piece == WHITE else -1
        moves = []
        # ход на одну клетку
        for dc in (-1, 1):
            nr, nc = row + direction, col + dc
            if self.is_valid_pos(nr, nc) and self.board[nr][nc] == EMPTY:
                moves.append(((row, col), (nr, nc), None))
        # захват (прыжок)
        for dc in (-1, 1):
            nr, nc = row + direction*2, col + dc*2
            if self.is_valid_pos(nr, nc):
                mid_r, mid_c = row + direction, col + dc
                if self.is_valid_pos(mid_r, mid_c) and self.board[mid_r][mid_c] != EMPTY and self.board[mid_r][mid_c] != piece:
                    if self.board[nr][nc] == EMPTY:
                        moves.append(((row, col), (nr, nc), (mid_r, mid_c)))
        return moves

    def get_all_moves(self, color):
        moves = []
        for row in range(8):
            for col in range(8):
                if self.board[row][col] == color:
                    moves.extend(self.get_moves(row, col))
        return moves

    # Проверить, есть ли захваты
    def has_captures(self, color):
        for row in range(8):
            for col in range(8):
                if self.board[row][col] == color:
                    for move in self.get_moves(row, col):
                        if move[2] is not None:  # захват
                            return True
        return False

    def make_move(self, from_pos, to_pos, capture_pos=None):
        fr, fc = from_pos
        tr, tc = to_pos
        self.board[tr][tc] = self.board[fr][fc]
        self.board[fr][fc] = EMPTY
        if capture_pos:
            cr, cc = capture_pos
            self.board[cr][cc] = EMPTY

    def is_game_over(self):
        # если у одного из игроков нет шашек или нет ходов
        white_moves = self.get_all_moves(WHITE)
        black_moves = self.get_all_moves(BLACK)
        if not white_moves or not black_moves:
            return True
        # проверка на наличие шашек
        white_pieces = any(self.board[r][c] == WHITE for r in range(8) for c in range(8))
        black_pieces = any(self.board[r][c] == BLACK for r in range(8) for c in range(8))
        return not white_pieces or not black_pieces

    def get_winner(self):
        white_moves = self.get_all_moves(WHITE)
        black_moves = self.get_all_moves(BLACK)
        white_pieces = any(self.board[r][c] == WHITE for r in range(8) for c in range(8))
        black_pieces = any(self.board[r][c] == BLACK for r in range(8) for c in range(8))
        if not white_pieces or not white_moves:
            return BLACK
        if not black_pieces or not black_moves:
            return WHITE
        return None

    def parse_move(self, s):
        # ожидается "a2 b3"
        parts = s.strip().split()
        if len(parts) != 2:
            return None
        try:
            col1 = ord(parts[0][0]) - ord('a')
            row1 = int(parts[0][1]) - 1
            col2 = ord(parts[1][0]) - ord('a')
            row2 = int(parts[1][1]) - 1
            if not (0 <= row1 < 8 and 0 <= col1 < 8 and 0 <= row2 < 8 and 0 <= col2 < 8):
                return None
            return (row1, col1), (row2, col2)
        except:
            return None

    def play(self):
        print("Добро пожаловать в мини-шашки!")
        print("Белые (○) ходят первыми.")
        print("Вводите ход в формате: a2 b3")
        while not self.is_game_over():
            self.print_board()
            color = self.turn
            color_name = "белых (○)" if color == WHITE else "чёрных (●)"
            print(f"Ход {color_name}.")
            # Проверяем, есть ли обязательные захваты
            captures = self.has_captures(color)
            # Получаем все возможные ходы
            all_moves = self.get_all_moves(color)
            if not all_moves:
                print("Нет доступных ходов, игра окончена.")
                break
            # Фильтруем ходы, если есть захваты – только захваты
            if captures:
                moves = [m for m in all_moves if m[2] is not None]
                print("Обязательный захват!")
            else:
                moves = all_moves
            # Ввод от пользователя
            while True:
                cmd = input("> ").strip()
                if cmd.lower() == 'quit':
                    return
                parsed = self.parse_move(cmd)
                if parsed is None:
                    print("Неверный формат. Используйте: a2 b3")
                    continue
                from_pos, to_pos = parsed
                # Ищем ход в списке
                found = None
                for m in moves:
                    if m[0] == from_pos and m[1] == to_pos:
                        found = m
                        break
                if found is None:
                    print("Неверный ход. Попробуйте снова.")
                    continue
                # Выполняем ход
                self.make_move(found[0], found[1], found[2])
                self.turn = BLACK if color == WHITE else WHITE
                break
        # Конец игры
        self.print_board()
        winner = self.get_winner()
        if winner == WHITE:
            print("Победили белые (○)!")
        elif winner == BLACK:
            print("Победили чёрные (●)!")
        else:
            print("Ничья? (такого не должно быть)")

if __name__ == "__main__":
    game = Checkers()
    game.play()
